utils/zerovec/src/zerovec/mod.rs

```diff
@@ -0,0 +1,500 @@
+        use ZeroVec::*;
+        match self {
+            Owned(_) => write!(f, "ZeroVec::Owned({:?})", self.to_vec()),
+            Borrowed(_) => write!(f, "ZeroVec::Borrowed({:?})", self.to_vec()),
+        }
+    }
+}
+
+impl<'a, 'b, T> PartialEq<ZeroVec<'b, T>> for ZeroVec<'a, T>
+where
+    T: AsULE + Copy + PartialEq,
+{
+    #[inline(always)]
+    fn eq(&self, other: &ZeroVec<'b, T>) -> bool {
+        // Note: T implements PartialEq but not T::ULE
+        self.iter().eq(other.iter())
+    }
+}
+
+impl<T> PartialEq<&[T]> for ZeroVec<'_, T>
+where
+    T: AsULE + Copy + PartialEq,
+{
+    #[inline(always)]
+    fn eq(&self, other: &&[T]) -> bool {
+        self.iter().eq(other.iter().copied())
+    }
+}
+
+impl<'a, T> ZeroVec<'a, T>
+where
+    T: AsULE + ?Sized,
+{
+    /// Parses a `&[u8]` buffer into a `ZeroVec<T>`.
+    ///
+    /// This function is infallible for built-in integer types, but fallible for other types,
+    /// such as `char`. For more information, see [`ULE::parse_byte_slice`].
+    ///
+    /// The bytes within the byte buffer must remain constant for the life of the ZeroVec.
+    ///
+    /// # Endianness
+    ///
+    /// The byte buffer must be encoded in little-endian, even if running in a big-endian
+    /// environment. This ensures a consistent representation of data across platforms.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use zerovec::ZeroVec;
+    ///
+    /// let bytes: &[u8] = &[0xD3, 0x00, 0x19, 0x01, 0xA5, 0x01, 0xCD, 0x01];
+    /// let zerovec: ZeroVec<u16> = ZeroVec::try_from_bytes(bytes).expect("infallible");
+    ///
+    /// assert!(matches!(zerovec, ZeroVec::Borrowed(_)));
+    /// assert_eq!(zerovec.get(2), Some(421));
+    /// ```
+    pub fn try_from_bytes(bytes: &'a [u8]) -> Result<Self, <<T as AsULE>::ULE as ULE>::Error> {
+        let slice: &'a [T::ULE] = T::ULE::parse_byte_slice(bytes)?;
+        Ok(Self::Borrowed(slice))
+    }
+
+    /// Returns a `ZeroVec<T>` as its underlying `&[u8]` byte buffer representation.
+    ///
+    /// Useful for serialization.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use zerovec::ZeroVec;
+    ///
+    /// // The little-endian bytes correspond to the numbers on the following line.
+    /// let bytes: &[u8] = &[0xD3, 0x00, 0x19, 0x01, 0xA5, 0x01, 0xCD, 0x01];
+    /// let nums: &[u16] = &[211, 281, 421, 461];
+    ///
+    /// let zerovec = ZeroVec::from_aligned(nums);
+    ///
+    /// assert_eq!(bytes, zerovec.as_bytes());
+    /// ```
+    #[inline(always)]
+    pub fn as_bytes(&self) -> &[u8] {
+        T::ULE::as_byte_slice(self.as_slice())
+    }
+
+    /// Dereferences this `ZeroVec<T>` as `&[T::ULE]`. Most other functions on `ZeroVec<T>` use
+    /// this function as a building block.
+    #[inline(always)]
+    pub fn as_slice(&self) -> &[T::ULE] {
+        use ZeroVec::*;
+        match self {
+            Owned(vec) => vec.as_slice(),
+            Borrowed(slice) => slice,
```

Comment:
`suggestion(not-blocking)`: instead of `use ZeroVec::*` you could do `match self { Self::Owned, ... }` which I think is much more readable.